### PR TITLE
Spice mnist

### PR DESCRIPTION
**Pull request by rhendz**
added agent logic to support training, testing, and validation for spice mnist model

### CODE REVIEW

File: spice/training/actions.py [hunk 1]
```diff
@@ -516,20 +522,75 @@ def tokenize_function(examples):
 
         return tokenized_dataset
 
+    def _process_dataset(self, dataset, config, task):
+        # get image processor
+        print("Loading image processor...")
```
Comment (djstein) [inline spice/training/actions.py]:
```suggestion
        LOGGGER.info("Loading image processor...")
```

File: spice/training/actions.py [hunk 2]
```diff
@@ -516,20 +522,75 @@ def tokenize_function(examples):
 
         return tokenized_dataset
 
+    def _process_dataset(self, dataset, config, task):
+        # get image processor
+        LOGGGER.info("Loading image processor...")
+        if task == "train" or task == "test":
```
Comment (djstein) [inline spice/training/actions.py]:
```suggestion
        if task in ["train", "test"]:
```

File: spice/training/actions.py [hunk 3]
```diff
@@ -516,20 +522,75 @@ def tokenize_function(examples):
...
+
+            image_processor = AutoImageProcessor.from_pretrained(
+                pretrained_model_name_or_path=hf_model_repo_id
+                if hf_model_repo_id and training_round_number > 1
+                else base_model_repo_id,
+                revision="main"  # uses main tokenizer
+                if hf_model_repo_id and training_round_number > 1
+                else base_model_repo_revision,
+                trust_remote_code=True,
+            )
```
Comment (djstein) [inline spice/training/actions.py]: pretty verbose, lets move the ternaries out
```suggestion
            pretrained_model_name_or_path = base_model_repo_id
            revision = "main"
            if hf_model_repo_id and training_round_number > 1:
                pretrained_model_name_or_path = base_model_repo_id
                revision = base_model_repo_revision

            image_processor = AutoImageProcessor.from_pretrained(
                pretrained_model_name_or_path= pretrained_model_name_or_path,
                revision=revison,
                trust_remote_code=True,
            )
```

File: spice/training/actions.py [hunk 4]
```diff
@@ -516,20 +522,75 @@ def tokenize_function(examples):
...
+            raise ValueError(error_message)
+
+        normalize = Normalize(
+            mean=image_processor.image_mean, std=image_processor.image_std
+        )
+        size = (
+            image_processor.size["shortest_edge"]
+            if "shortest_edge" in image_processor.size
+            else (image_processor.size["height"], image_processor.size["width"])
+        )
```
Comment (djstein) [inline spice/training/actions.py]: 4 potential exceptions here. we want to validate that the dict get's are actually there or we can end up with a size equal to None

we might want to do:
```suggestion
size = image_processor.size.get("shortest_edge", None)
if not size and image_processor.size.get("height", None) and image_processor.size.get("width", None):
    size = (image_processor.size["height"], image_processor.size["width"])
else:
    raise Exception("size was not set. image_processor did not have keys shortest_edge, or height and width.")
```

Comment (djstein) [reply spice/training/actions.py]: is there also a case where height and width can be 0?

File: spice/training/actions.py [hunk 5]
```diff
@@ -516,20 +522,75 @@ def tokenize_function(examples):
...
+        _transforms = Compose([RandomResizedCrop(size), ToTensor(), normalize])
```
Comment (djstein) [inline spice/training/actions.py:576]: if `_transforms` and `normalize` are only used in the `transforms` function should we move them into it?

File: spice/training/actions.py [hunk 6]
```diff
@@ -549,11 +611,24 @@ def _load_model(self, config, task):
             LOGGER.error(error_message)
             raise ValueError(error_message)
 
-        # load your model with the number of expected labels:
+        # Load your model with the number of expected labels:
         print(f"Loading {task} model...")
```
Comment (djstein) [inline spice/training/actions.py]:
```suggestion
        LOGGER.info(f"Loading {task} model...")
```

File: spice/training/actions.py [hunk 7]
```diff
@@ -603,7 +679,19 @@ def train_model(self):
...
+                train_dataset, config, "train"
+            )
+        elif base_model_repo_id == "bert-base-uncased":
+            tokenized_dataset = self._tokenize_dataset(train_dataset, config, "train")
+        else:
+            error_message = (
+                f"train_model has unknown base model: {base_model_repo_id} type"
+            )
+            LOGGER.error(error_message)
+            raise ValueError(error_message)
```
Comment (djstein) [inline spice/training/actions.py:705]: we'll definitely need to to change this to be based on some attribute of the training round like "model type" of "image" or "text", etc

**Review by djstein — CHANGES_REQUESTED** [review]
please review comments

**Review by djstein — APPROVED** [review]
